python/dgl/contrib/sampling/sampler.py

```diff
@@ -77,10 +84,124 @@ def __next__(self):
+                pass
+            except RuntimeError as e:
+                tb = traceback.format_exc()
+                self._errorq.put((e, tb))
+                self._dataq.put(None)
+
+            try:
+                data = next(loader_iter)
+                error = None
+            except Exception as e:  # pylint: disable=broad-except
+                tb = traceback.format_exc()
+                error = (e, tb)
+                data = None
+            finally:
+                self._errorq.put(error)
+                self._dataq.put(data)
+
+    def __next__(self):
+        next_item = self._dataq.get()
+        next_error = self._errorq.get()
+
+        if next_error is None:
+            return next_item
+        else:
+            self._controlq.put(None)
+            if isinstance(next_error[0], StopIteration):
+                raise StopIteration
+            else:
+                return self._reraise(*next_error)
+
+    def _reraise(self, e, tb):
+        print('Reraising exception from Prefetcher', file=sys.stderr)
+        print(tb, file=sys.stderr)
+        raise e
+
+    def _check_start(self):
+        assert not self._checked_start
+        self._checked_start = True
+        next_error = self._errorq.get(block=True)
+        if next_error is not None:
+            self._reraise(*next_error)
+
+    def next(self):
+        return self.__next__()
+
+
+class _ThreadPrefetcher(_Prefetcher, threading.Thread):
+    """Internal threaded prefetcher."""
+
+    def __init__(self, *args, **kwargs):
+        super(_ThreadPrefetcher, self).__init__(*args, **kwargs)
+        self._dataq = queue.Queue(self.num_prefetch)
+        self._controlq = queue.Queue()
+        self._errorq = queue.Queue(self.num_prefetch)
+        self.daemon = True
+        self.start()
+        self._check_start()
+
+class _PrefetchingLoader(object):
+    """Prefetcher for a Loader in a separate Thread or Process.
+    This iterator will create another thread or process to perform
+    ``iter_next`` and then store the data in memory. It potentially accelerates
+    the data read, at the cost of more memory usage.
+
+    Parameters
+    ----------
+    loader : an iterator
+        Source loader.
+    num_prefetch : int, default 1
+        Number of elements to prefetch from the loader. Must be greater 0.
+    """
+
+    def __init__(self, loader, num_prefetch=1):
+        self._loader = loader
+        self._num_prefetch = num_prefetch
+        if num_prefetch < 1:
+            raise ValueError('num_prefetch must be greater 0.')
+
+    def __iter__(self):
+        return _ThreadPrefetcher(self._loader, self._num_prefetch)
+
 def NeighborSampler(g, batch_size, expand_factor, num_hops=1,
                     neighbor_type='in', node_prob=None, seed_nodes=None,
                     shuffle=False, num_workers=1, max_subgraph_size=None,
-                    return_seed_id=False):
+                    return_seed_id=False, prefetch=False):
```

Comment:
prefetch might be misleading. when num_workers > 1, we already compute some subgraphs in advance.